### DISCUSSION TITLE
Gradle test delegation

### DISCUSSION THREAD

**Discussion by kylechui** [mfussenegger/nvim-jdtls#729] [General]
I've recently been running into issues in an older codebase where the testing platform from vscode-java-test seems to be incompatible with my actual tests---in particular, the debugger will fail to start. This happened after I updated vscode-java-test, but now I am struggling to downgrade as a lot of the dependencies it pulls in are now dead links. Is there any way to delegate the test running to gradle itself, e.g. in a similar way described here? https://github.com/microsoft/vscode-java-test/issues/1045

**Comment by mfussenegger** [comment]
There's currently no direct support for that. From a quick look at vscode-gradle they're running gradle with `-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=localhost:%d", javaDebugPort)` and then connect to that port but there's quite a bit of glue code surrounding that.

But if it worked before and now doesn't I'd think it is a regression in vscode-java-test that should be reported?

**Reply by kylechui** [reply]
I see, thanks for the response! Somewhat related to the debug port---the nvim-jdtls plugin is outputting "Could not resolve classpaths. Project may have compile errors or unresolved dependencies" (which I don't think is the case). The Neovim error trace then says

```
...dap.lua:1191: Adapter used with attach must have a port property
... in function 'assert'
```

Is there any more insight that can be gleaned from this before I go ask the vscode-java-test devs? Thanks again!

**Reply by mfussenegger** [reply]
> ...dap.lua:1191: Adapter used with attach must have a port property.. in function 'assert'

is likely a follow up error due to "Could not resolve classpaths. Project may have compile errors or unresolved dependencies"

nvim-jdtls calls the `vscode.java.resolveMainClass` command whenever a debug session starts where the configuration doesn't already have mainClass, projectName, modulePaths, classPaths and javaExec set.

If `:JdtShowLogs` doesn't provide more information the vscode-java-test maintainers are likely going to need a sample project that reproduces the issue.
